modules/git/hash.go

```diff
@@ -0,0 +1,239 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package git
+
+import (
+	"bytes"
+	"crypto/sha1"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"hash"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+type ObjectFormatID int
+
+const (
+	Sha1 ObjectFormatID = iota
+)
+
+// sha1Pattern can be used to determine if a string is an valid sha
+var sha1Pattern = regexp.MustCompile(`^[0-9a-f]{4,40}$`)
+
+type ObjectID interface {
+	String() string
+	IsZero() bool
+	RawValue() []byte
+	Type() ObjectFormat
+}
+
+type ObjectFormat interface {
+	ID() ObjectFormatID
+	String() string
+
+	// Empty is the hash of empty git
+	Empty() ObjectID
```

Comment:
Could we have an `IsObjectIDEmpty` function? So in many places, we can just use this function but not use an equal sign?

Reply:
There is already another helper function here, now in object_id.go `IsEmptyCommitID(string)`.

I didn't want to change the mechanic of having these string comparisons everywhere because we want a no-op patch here that should just refactor the hash function into one place. In future changes, I was hoping of refactoring other things, like all the commitID from a `string` into `commitID` (one structure at a time) which would address these problems of string comparisons in at same time. I'm afraid that if we start to change too many things in one patch, it will just get larger and larger and become like an even larger octopus :laughing: 

What do you think?